File: Algorithms/14-11-22/shell-sort.cpp
#include <iostream>
const int n=8;
void shellSort(int arr[], int size){
	for(int gap=size/2; gap>0; gap/=2){
		for(int i = gap; i<size; ++i){
			for(int j = i-gap; j>=0; j-=gap){
				if(arr[i]>=arr[j]){
					break;
				}
				else{
					std::swap(arr[i], arr[j]);
				}
			}
		}
	}
}


void insertArray(int arr[], int size){
	for(int i = 0; i<size; ++i){
		std::cout<<"Array["<<i<<"]="; 
		std::cin>>arr[i];
	}
}

int main(){
	int array[n];
	insertArray(array, n);
	shellSort(array, n);

	std::cout<<"After shell sort: \n";
	for(int i = 0; i<n; ++i){
	std::cout<<"Array["<<i<<"]="<<array[i]<<"\n";
	}
	return 0;
}

File: exam/python/shell-sort.py
llist =[9,8,3,7,5,6,4,1]

def ShellSort(llist):
    N = len(llist)
    interval = N//2
    while interval>0:
        for i in range(interval, N):
            key = llist[i]
            j = i-interval
            while key<llist[j] and j>=0:
                llist[j+interval] = llist[j]
                j-=interval

            llist[j+interval] = key
        interval//=2

    return llist

print("Shell sort: ", ShellSort(llist))
